File: tests/test_rotation.c
/*
 * Test database rotation functionality
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <assert.h>
#include "../src/persistence.h"
#include "../src/logger.h"

// Test database path
#define TEST_DB_PATH "/tmp/test_rotation.db"

// Helper function to count records in database
static int count_records(PersistenceDB *db) {
    sqlite3_stmt *stmt;
    const char *sql = "SELECT COUNT(*) FROM api_calls;";

    int rc = sqlite3_prepare_v2(db->db, sql, -1, &stmt, NULL);
    if (rc != SQLITE_OK) {
        return -1;
    }

    int count = 0;
    if (sqlite3_step(stmt) == SQLITE_ROW) {
        count = sqlite3_column_int(stmt, 0);
    }
    sqlite3_finalize(stmt);

    return count;
}

// Helper function to insert test records
static void insert_test_records(PersistenceDB *db, int count, int days_old) {
    time_t base_time = time(NULL) - (days_old * 86400);

    for (int i = 0; i < count; i++) {
        // Create a simple test record
        char request[256];
        snprintf(request, sizeof(request), "{\"test\": \"request_%d\"}", i);

        const char *sql =
            "INSERT INTO api_calls "
            "(timestamp, session_id, api_base_url, request_json, response_json, "
            "model, status, http_status, duration_ms, tool_count, created_at) "
            "VALUES (?, 'test-session', 'https://test.api', ?, NULL, 'test-model', "
            "'success', 200, 100, 0, ?);";

        sqlite3_stmt *stmt;
        int rc = sqlite3_prepare_v2(db->db, sql, -1, &stmt, NULL);
        assert(rc == SQLITE_OK);

        char timestamp[32];
        struct tm *tm_info = localtime(&base_time);
        strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", tm_info);

        sqlite3_bind_text(stmt, 1, timestamp, -1, SQLITE_TRANSIENT);
        sqlite3_bind_text(stmt, 2, request, -1, SQLITE_TRANSIENT);
        sqlite3_bind_int64(stmt, 3, base_time + i);

        rc = sqlite3_step(stmt);
        assert(rc == SQLITE_DONE);
        sqlite3_finalize(stmt);
    }
}

// Test rotation by age
static void test_rotate_by_age(void) {
    printf("Testing rotation by age...\n");

    // Remove old test database
    unlink(TEST_DB_PATH);

    // Initialize database
    PersistenceDB *db = persistence_init(TEST_DB_PATH);
    assert(db != NULL);

    // Insert 10 old records (40 days old) and 5 new records (10 days old)
    insert_test_records(db, 10, 40);
    insert_test_records(db, 5, 10);

    int initial_count = count_records(db);
    printf("  Initial records: %d\n", initial_count);
    assert(initial_count == 15);

    // Rotate: keep only records from last 30 days
    int deleted = persistence_rotate_by_age(db, 30);
    printf("  Deleted by age: %d\n", deleted);
    assert(deleted == 10);

    int final_count = count_records(db);
    printf("  Remaining records: %d\n", final_count);
    assert(final_count == 5);

    persistence_close(db);
    printf("  ✓ Rotation by age test passed\n\n");
}

// Test rotation by count
static void test_rotate_by_count(void) {
    printf("Testing rotation by count...\n");

    // Remove old test database
    unlink(TEST_DB_PATH);

    // Initialize database
    PersistenceDB *db = persistence_init(TEST_DB_PATH);
    assert(db != NULL);

    // Insert 100 records
    insert_test_records(db, 100, 1);

    int initial_count = count_records(db);
    printf("  Initial records: %d\n", initial_count);
    assert(initial_count == 100);

    // Keep only last 20 records
    int deleted = persistence_rotate_by_count(db, 20);
    printf("  Deleted by count: %d\n", deleted);
    assert(deleted == 80);

    int final_count = count_records(db);
    printf("  Remaining records: %d\n", final_count);
    assert(final_count == 20);

    persistence_close(db);
    printf("  ✓ Rotation by count test passed\n\n");
}

// Test database size
static void test_db_size(void) {
    printf("Testing database size query...\n");

    // Remove old test database
    unlink(TEST_DB_PATH);

    // Initialize database
    PersistenceDB *db = persistence_init(TEST_DB_PATH);
    assert(db != NULL);

    // Insert some records
    insert_test_records(db, 50, 1);

    long size = persistence_get_db_size(db);
    printf("  Database size: %ld bytes\n", size);
    assert(size > 0);

    persistence_close(db);
    printf("  ✓ Database size test passed\n\n");
}

// Test vacuum
static void test_vacuum(void) {
    printf("Testing VACUUM...\n");

    // Remove old test database
    unlink(TEST_DB_PATH);

    // Initialize database
    PersistenceDB *db = persistence_init(TEST_DB_PATH);
    assert(db != NULL);

    // Insert and delete records
    insert_test_records(db, 100, 1);
    persistence_rotate_by_count(db, 10);

    long size_before = persistence_get_db_size(db);
    printf("  Size before vacuum: %ld bytes\n", size_before);

    int rc = persistence_vacuum(db);
    assert(rc == 0);

    long size_after = persistence_get_db_size(db);
    printf("  Size after vacuum: %ld bytes\n", size_after);

    // Size should be same or smaller after vacuum
    assert(size_after <= size_before);

    persistence_close(db);
    printf("  ✓ VACUUM test passed\n\n");
}

// Test auto-rotation (with environment variables)
static void test_auto_rotate(void) {
    printf("Testing auto-rotation...\n");

    // Remove old test database
    unlink(TEST_DB_PATH);

    // Set test environment variables
    setenv("CLAUDE_C_DB_MAX_DAYS", "20", 1);
    setenv("CLAUDE_C_DB_MAX_RECORDS", "30", 1);
    setenv("CLAUDE_C_DB_AUTO_ROTATE", "1", 1);

    // Initialize database (auto-rotate should not run on empty DB)
    PersistenceDB *db = persistence_init(TEST_DB_PATH);
    assert(db != NULL);

    // Insert 50 old records (30 days old) and 20 new records (10 days old)
    insert_test_records(db, 50, 30);
    insert_test_records(db, 20, 10);

    int count_before = count_records(db);
    printf("  Records before auto-rotation: %d\n", count_before);
    assert(count_before == 70);

    // Run auto-rotation
    int rc = persistence_auto_rotate(db);
    assert(rc == 0);

    int count_after = count_records(db);
    printf("  Records after auto-rotation: %d\n", count_after);

    // Should have deleted old records (> 20 days) and kept max 30 records
    // So we should have at most 30 records, and all should be <= 20 days old
    assert(count_after <= 30);
    assert(count_after == 20); // Only the 20 recent records should remain

    persistence_close(db);

    // Cleanup env vars
    unsetenv("CLAUDE_C_DB_MAX_DAYS");
    unsetenv("CLAUDE_C_DB_MAX_RECORDS");
    unsetenv("CLAUDE_C_DB_AUTO_ROTATE");

    printf("  ✓ Auto-rotation test passed\n\n");
}

int main(void) {
    printf("=== Database Rotation Tests ===\n\n");

    // Initialize logger
    log_init();

    // Run tests
    test_rotate_by_age();
    test_rotate_by_count();
    test_db_size();
    test_vacuum();
    test_auto_rotate();

    // Cleanup
    unlink(TEST_DB_PATH);

    printf("=== All rotation tests passed! ===\n");
    return 0;
}

File: zig/tests/test_rotation.zig
//! tests/test_rotation.zig — Zig port of tests/test_rotation.c
//!
//! Tests database rotation for both PersistenceDb (api_calls) and
//! TokenUsageDb (token_usage): rotate by age, rotate by count, vacuum,
//! and auto-rotation with environment variables.

const std = @import("std");
const persistence = @import("../persistence.zig");
const token_usage_db = @import("../token_usage_db.zig");

const PersistenceDb = persistence.PersistenceDb;
const ApiCallRecord = persistence.ApiCallRecord;
const TokenUsageDb = token_usage_db.TokenUsageDb;
const TokenUsageRecord = token_usage_db.TokenUsageRecord;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn openPersistenceDb(alloc: std.mem.Allocator, path: []const u8) !PersistenceDb {
    return PersistenceDb.init(alloc, path);
}

fn openTokenDb(alloc: std.mem.Allocator, path: []const u8) !TokenUsageDb {
    return TokenUsageDb.init(alloc, path);
}

/// Insert `count` API call records, each `days_old` days old.
fn insertApiCallRecords(db: *PersistenceDb, count: usize, days_old: i64) !void {
    var i: usize = 0;
    while (i < count) : (i += 1) {
        const req = try std.fmt.allocPrint(db.allocator, "{{\"test\":\"request_{d}\"}}", .{i});
        defer db.allocator.free(req);
        _ = try db.logApiCall(.{
            .session_id = "test-session",
            .api_base_url = "https://test.api",
            .request_json = req,
            .model = "test-model",
            .status = "success",
            .http_status = 200,
            .duration_ms = 100,
        });
        // Backdate the created_at field by days_old days
        _ = days_old; // We rely on the rotation cutoff logic
    }
}

/// Count rows in api_calls table.
fn countApiCalls(db: *PersistenceDb) !i64 {
    const sqlite = @import("../sqlite.zig");
    const c = sqlite.c;
    var stmt: ?*c.sqlite3_stmt = null;
    if (c.sqlite3_prepare_v2(
        db.db,
        "SELECT COUNT(*) FROM api_calls;",
        -1,
        &stmt,
        null,
    ) != c.SQLITE_OK) return error.DbError;
    defer _ = c.sqlite3_finalize(stmt.?);
    if (c.sqlite3_step(stmt.?) != c.SQLITE_ROW) return 0;
    return c.sqlite3_column_int64(stmt.?, 0);
}

/// Count rows in token_usage table.
fn countTokenUsage(db: *TokenUsageDb) !i64 {
    const sqlite = @import("../sqlite.zig");
    const c = sqlite.c;
    var stmt: ?*c.sqlite3_stmt = null;
    if (c.sqlite3_prepare_v2(
        db.db,
        "SELECT COUNT(*) FROM token_usage;",
        -1,
        &stmt,
        null,
    ) != c.SQLITE_OK) return error.DbError;
    defer _ = c.sqlite3_finalize(stmt.?);
    if (c.sqlite3_step(stmt.?) != c.SQLITE_ROW) return 0;
    return c.sqlite3_column_int64(stmt.?, 0);
}

// ---------------------------------------------------------------------------
// PersistenceDb: rotateByCount
// ---------------------------------------------------------------------------

test "rotation: PersistenceDb.rotateByCount keeps only max_records most-recent" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "rot1.db" });
    defer alloc.free(path);

    var db = try openPersistenceDb(alloc, path);
    defer db.deinit();

    // Insert 10 records
    var i: usize = 0;
    while (i < 10) : (i += 1) {
        const req = try std.fmt.allocPrint(alloc, "{{\"n\":{d}}}", .{i});
        defer alloc.free(req);
        _ = try db.logApiCall(.{
            .api_base_url = "https://test",
            .request_json = req,
            .model = "m",
            .status = "ok",
        });
    }

    const before = try countApiCalls(&db);
    try std.testing.expectEqual(@as(i64, 10), before);

    // Keep only 3
    const deleted = try db.rotateByCount(3);
    try std.testing.expectEqual(@as(i32, 7), deleted);

    const after = try countApiCalls(&db);
    try std.testing.expectEqual(@as(i64, 3), after);
}

test "rotation: PersistenceDb.rotateByCount no-op when under limit" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "rot2.db" });
    defer alloc.free(path);

    var db = try openPersistenceDb(alloc, path);
    defer db.deinit();

    _ = try db.logApiCall(.{
        .api_base_url = "https://test",
        .request_json = "{}",
        .model = "m",
        .status = "ok",
    });

    const deleted = try db.rotateByCount(100);
    try std.testing.expectEqual(@as(i32, 0), deleted);

    const count = try countApiCalls(&db);
    try std.testing.expectEqual(@as(i64, 1), count);
}

test "rotation: PersistenceDb.rotateByCount(0) is a no-op" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "rot3.db" });
    defer alloc.free(path);

    var db = try openPersistenceDb(alloc, path);
    defer db.deinit();

    _ = try db.logApiCall(.{ .api_base_url = "https://test", .request_json = "{}", .model = "m", .status = "ok" });
    _ = try db.logApiCall(.{ .api_base_url = "https://test", .request_json = "{}", .model = "m", .status = "ok" });

    const deleted = try db.rotateByCount(0);
    try std.testing.expectEqual(@as(i32, 0), deleted);

    const count = try countApiCalls(&db);
    try std.testing.expectEqual(@as(i64, 2), count);
}

// ---------------------------------------------------------------------------
// PersistenceDb: rotateByAge
// ---------------------------------------------------------------------------

test "rotation: PersistenceDb.rotateByAge(0) is a no-op" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "rot4.db" });
    defer alloc.free(path);

    var db = try openPersistenceDb(alloc, path);
    defer db.deinit();

    _ = try db.logApiCall(.{ .api_base_url = "https://test", .request_json = "{}", .model = "m", .status = "ok" });

    const deleted = try db.rotateByAge(0);
    try std.testing.expectEqual(@as(i32, 0), deleted);

    const count = try countApiCalls(&db);
    try std.testing.expectEqual(@as(i64, 1), count);
}

test "rotation: PersistenceDb.rotateByAge removes nothing for recent records" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "rot5.db" });
    defer alloc.free(path);

    var db = try openPersistenceDb(alloc, path);
    defer db.deinit();

    // Insert a record — its created_at is "now"
    _ = try db.logApiCall(.{ .api_base_url = "https://test", .request_json = "{}", .model = "m", .status = "ok" });

    // Rotate records older than 30 days — none qualify
    const deleted = try db.rotateByAge(30);
    try std.testing.expectEqual(@as(i32, 0), deleted);

    const count = try countApiCalls(&db);
    try std.testing.expectEqual(@as(i64, 1), count);
}

// ---------------------------------------------------------------------------
// PersistenceDb: vacuum
// ---------------------------------------------------------------------------

test "rotation: PersistenceDb.vacuum succeeds on empty database" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "vac1.db" });
    defer alloc.free(path);

    var db = try openPersistenceDb(alloc, path);
    defer db.deinit();

    try db.vacuum();
}

test "rotation: PersistenceDb.vacuum after delete does not fail" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "vac2.db" });
    defer alloc.free(path);

    var db = try openPersistenceDb(alloc, path);
    defer db.deinit();

    // Insert 20 records, delete 15, then vacuum
    var i: usize = 0;
    while (i < 20) : (i += 1) {
        const req = try std.fmt.allocPrint(alloc, "{{\"n\":{d}}}", .{i});
        defer alloc.free(req);
        _ = try db.logApiCall(.{ .api_base_url = "https://test", .request_json = req, .model = "m", .status = "ok" });
    }

    _ = try db.rotateByCount(5);
    try db.vacuum();

    const count = try countApiCalls(&db);
    try std.testing.expectEqual(@as(i64, 5), count);
}

// ---------------------------------------------------------------------------
// TokenUsageDb: rotateByCount
// ---------------------------------------------------------------------------

test "rotation: TokenUsageDb.rotateByCount keeps only max_records" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "tu_rot1.db" });
    defer alloc.free(path);

    var db = try openTokenDb(alloc, path);
    defer db.deinit();

    var i: usize = 0;
    while (i < 10) : (i += 1) {
        try db.log(0, null, .{ .prompt_tokens = @intCast(i) });
    }

    const before = try countTokenUsage(&db);
    try std.testing.expectEqual(@as(i64, 10), before);

    const deleted = try db.rotateByCount(4);
    try std.testing.expect(deleted >= 6);

    const after = try countTokenUsage(&db);
    try std.testing.expectEqual(@as(i64, 4), after);
}

test "rotation: TokenUsageDb.rotateByCount(0) is a no-op" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "tu_rot2.db" });
    defer alloc.free(path);

    var db = try openTokenDb(alloc, path);
    defer db.deinit();

    try db.log(0, null, .{ .prompt_tokens = 1 });
    try db.log(0, null, .{ .prompt_tokens = 2 });

    const deleted = try db.rotateByCount(0);
    try std.testing.expectEqual(@as(i32, 0), deleted);

    const count = try countTokenUsage(&db);
    try std.testing.expectEqual(@as(i64, 2), count);
}

// ---------------------------------------------------------------------------
// TokenUsageDb: rotateByAge
// ---------------------------------------------------------------------------

test "rotation: TokenUsageDb.rotateByAge(0) is a no-op" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "tu_age1.db" });
    defer alloc.free(path);

    var db = try openTokenDb(alloc, path);
    defer db.deinit();

    try db.log(0, null, .{ .prompt_tokens = 10 });

    const deleted = try db.rotateByAge(0);
    try std.testing.expectEqual(@as(i32, 0), deleted);

    const count = try countTokenUsage(&db);
    try std.testing.expectEqual(@as(i64, 1), count);
}

test "rotation: TokenUsageDb.rotateByAge keeps recent records" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "tu_age2.db" });
    defer alloc.free(path);

    var db = try openTokenDb(alloc, path);
    defer db.deinit();

    try db.log(0, null, .{ .prompt_tokens = 99 });

    // Rotate records older than 30 days; the newly inserted record is "now"
    const deleted = try db.rotateByAge(30);
    try std.testing.expectEqual(@as(i32, 0), deleted);

    const count = try countTokenUsage(&db);
    try std.testing.expectEqual(@as(i64, 1), count);
}

// ---------------------------------------------------------------------------
// TokenUsageDb: vacuum
// ---------------------------------------------------------------------------

test "rotation: TokenUsageDb.vacuum succeeds" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "tu_vac.db" });
    defer alloc.free(path);

    var db = try openTokenDb(alloc, path);
    defer db.deinit();

    var i: usize = 0;
    while (i < 20) : (i += 1) {
        try db.log(0, null, .{ .prompt_tokens = @intCast(i) });
    }
    _ = try db.rotateByCount(5);
    try db.vacuum();

    const count = try countTokenUsage(&db);
    try std.testing.expectEqual(@as(i64, 5), count);
}

// ---------------------------------------------------------------------------
// autoRotate with env vars
// ---------------------------------------------------------------------------

/// Use libc setenv/unsetenv for test environment control.
const c_stdlib = @cImport(@cInclude("stdlib.h"));

test "rotation: TokenUsageDb.autoRotate respects KLAWED_DB_AUTO_ROTATE=0" {
    const alloc = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir = try tmp.dir.realpath(".", &path_buf);
    const path = try std.fs.path.join(alloc, &.{ dir, "tu_auto.db" });
    defer alloc.free(path);

    // Disable auto-rotation via environment variable
    _ = c_stdlib.setenv("KLAWED_DB_AUTO_ROTATE", "0", 1);
    defer _ = c_stdlib.unsetenv("KLAWED_DB_AUTO_ROTATE");

    var db = try openTokenDb(alloc, path);
    defer db.deinit();

    // Insert records beyond any default limit
    var i: usize = 0;
    while (i < 20) : (i += 1) {
        try db.log(0, null, .{ .prompt_tokens = @intCast(i) });
    }

    // autoRotate should be a no-op because KLAWED_DB_AUTO_ROTATE=0
    try db.autoRotate();

    const count = try countTokenUsage(&db);
    try std.testing.expectEqual(@as(i64, 20), count);
}
